### PR TITLE
Feature/protect file extensions

### PR DESCRIPTION
**Pull request by markadrake**
# Protect File Extensions

I deployed the Umbraco CMS to a secure enterprise environment and encountered an issue. Our security policies blocked the static asset bundles produced by Smidge. After troubleshooting, we discovered that network requests were being blocked because the _file extensions were unknown_. Smidge's file naming pattern changes the file extension. 

With this PR, I hope to remedy the above issue by _opting in_ through a configuration value `protectFileExtensions`.

Please review my PR and I appreciate any critical feedback. I am _not_ a C# developer.

1. Adds a new configuration option, `protectFileExtensions`. 
2. Requires opt-in, so no existing behavior should be modified.
3. After opt-in, the naming scheme will preserve the file extension (`.css` or `.js`).
4. Adds two additional tests for verifying the path URL when `protectFileExtensions` is set to `true`.

# Sample Settings

image

# Previously, without `protectFileExtensions`

image

# Now, with `protectFileExtensions`
 
image

### CODE REVIEW

**Comment by markadrake** [comment]
@Shazwazza,

I ran into various errors after running `Smide.Web`.

1. The application failed to start because of a missing directory: `~/Smidge/Static`.
2. `Startup.cs:197` adds a non-existent file to a bundle. The application handles this error and continues to run.

Please let me know if you would like me to check in the changes I made to remedy these issues. Respectively:

1. Created the missing folder.
2. Removed lines 197-200.

File: src/Smidge.Core/SmidgeConfig.cs [hunk 1]
```diff
@@ -34,6 +34,8 @@ public SmidgeConfig()
         public string Version => _config["version"] ?? "1";
 
         public string DataFolder => (_config["dataFolder"] ?? "Smidge").Replace('/', Path.DirectorySeparatorChar);
+
+		public bool ProtectFileExtensions => bool.Parse((_config["protectFileExtensions"] ?? "false"));
```
Comment (PhyxionNL) [inline src/Smidge.Core/SmidgeConfig.cs:38]: This name is odd to me as there's nothing to protect. I think KeepFileExtensions would be a better name.

File: src/Smidge.Core/CompositeFiles/DefaultUrlManager.cs [hunk 2]
```diff
@@ -160,8 +162,8 @@ public ParsedUrlPath ParsePath(string input)
         private string GetCompositeUrl(string fileKey, string fileExtension, string cacheBusterValue)
         {
             //Create a delimited URL query string
-
-            const string handler = "~/{0}/{1}{2}.v{3}";
+
```
Comment (PhyxionNL) [inline src/Smidge.Core/CompositeFiles/DefaultUrlManager.cs:165]: Undo.

File: src/Smidge.Core/CompositeFiles/DefaultUrlManager.cs [hunk 3]
```diff
@@ -171,4 +173,4 @@ private string GetCompositeUrl(string fileKey, string fileExtension, string cach
                     cacheBusterValue));
         }
     }
-}
\ No newline at end of file
+}
```
Comment (PhyxionNL) [inline src/Smidge.Core/CompositeFiles/DefaultUrlManager.cs:176]: Undo.

**Review by PhyxionNL — CHANGES_REQUESTED** [review]

File: src/Smidge.Core/SmidgeConfig.cs (hunk 1 above)
Comment (markadrake) [reply src/Smidge.Core/SmidgeConfig.cs]: @PhyxionNL I've renamed everything to `keepFileExtensions`, including the 2 new tests as `Keep_File_Extensions`.

I noticed the tabs vs.spaces on line 38. I'm going to save checking this in until I have addressed your related feedback. 

Thank you!

**Comment by markadrake** [comment]
Please note that `protectFileExtensions` has been renamed to `keepFileExtensions`. 

**appsettings.json**:

image

**Smidge.Tests**:

image

File: src/Smidge.Core/CompositeFiles/DefaultUrlManager.cs (hunk 3 above)
Comment (markadrake) [reply src/Smidge.Core/CompositeFiles/DefaultUrlManager.cs]: Removed the new line on 176. Had to do this outside of Visual Studio.

File: src/Smidge.Core/CompositeFiles/DefaultUrlManager.cs (hunk 2 above)
Comment (markadrake) [reply src/Smidge.Core/CompositeFiles/DefaultUrlManager.cs]: Removed the new line on 165. Had to do this outside of Visual Studio.

**Comment by markadrake** [comment]
@Shazwazza or @PhyxionNL, do either of you have time to review the changes in this PR? I'm happy to continue addressing any issues you discover. I apologize for my persistence with this issue. 

All the best,

**Comment by markadrake** [comment]
@Shazwazza if there is any feedback at all, I'm happy to try and implement the changes. Hope all is good sir!

**Comment by Shazwazza** [comment]
Sorry for the delay, having a look now

**Comment by Shazwazza** [comment]
Looks awesome, thanks so much 🎉 

I'll ship a release asap

**Comment by Shazwazza** [comment]
This has been pushed :) https://github.com/Shazwazza/Smidge/releases/tag/v4.2.0

**Pull request merged**